### PR TITLE
Change the Scaler to track {sub,}domains

### PR DESCRIPTION
**Pull request by evancharlton**
Instead of the Scaler tracking the d3 transforms to handle zooming and
panning, track the domains and subdomains for each item. This allows
sub-components to query the domain & subdomain for any given item
(series or collection), and render themselves accordingly.

d3 is still used for the event handling, but the zoom / pan logic has
been moved to be handled entirely inside of the ZoomRect component.

This change will enable multiple components to share responsibility for
adjusting zoom. One example of this could be a scatterplot which allows
zooming on the chart itself, but also allowing the axes to further
manipulate the scaling in either direction.

This change also ends the notion that the x-axis represents time, since
this has not been true since the introduction of scatterplots. Instead,
each series now has three possible axes, each with their own domains and
subdomains: x, y, and time.

### CODE REVIEW

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

**Comment by evancharlton** [comment]
Full disclosure: this does not handle touch events. I will continue working on that while this PR is under review -- I expect it to be fully-contained in `ZoomRect`.

**Comment by evancharlton** [comment]
Update: touch events (pan & zoom) are now working (and, as mentioned, are fully-contained in ZoomRect)

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

File: src/components/ContextChart/index.js [hunk 1]
```diff
@@ -134,29 +157,31 @@ class ContextChart extends Component {
...
+    {({
+      domainsByItemId,
+      subDomainsByItemId,
+      timeSubDomain,
+      timeDomain,
+      contextSeries,
+      xScalerFactory,
+      updateDomains,
+    }) => (
+      <SizeMe monitorWidth>
```
Comment (danlevings) [inline src/components/ContextChart/index.js]: How does scatterplots deal with dynamic widths/heights?

Comment (evancharlton) [reply src/components/ContextChart/index.js]: Like line charts, a `<Scatterplot>` will be the height of its container (axes included in the math).

File: src/components/Scatterplot/index.js [hunk 2]
```diff
@@ -102,6 +100,7 @@ const ScatterplotComponent = ({
             onClick={onClick}
             zoomMode={ZoomMode.BOTH}
             xScalerFactory={xScalerFactory}
+            zoomAxes={{ x: true, y: true }}
```
Comment (danlevings) [inline src/components/Scatterplot/index.js]: i guess this means we always have scatter plots being zoomable? Maybe this should be an option like it is in Line Charts

Comment (evancharlton) [reply src/components/Scatterplot/index.js]: Good call -- the `zoomable` property overrides this (as an implementation detail of `InteractionLayer`). I'll make a story to show this, and think about if the API for this should be cleaned up.

Comment (evancharlton) [reply src/components/Scatterplot/index.js]: Oops, I misread it. Turning off zoomable doesn't work -- I'll fix that!

File: src/components/ZoomRect/index.js [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    const [touch] = touches;
+    const newTouchPosition = {
+      time: touch.pageX,
+      x: touch.pageX,
+      y: touch.pageY,
+    };
+    const updates = {};
+    itemIds.forEach(itemId => {
+      updates[itemId] = {};
+      ['x', 'y', 'time'].filter(axis => zoomAxes[axis]).forEach(axis => {
```
Comment (danlevings) [inline src/components/ZoomRect/index.js]: I see `x`, `y` and `time` referenced around a lot - should these be moved to something like `AXIS_TYPES`?

Comment (evancharlton) [reply src/components/ZoomRect/index.js]: Yeah, they definitely should be -- will do!

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

File: src/components/ZoomRect/index.js (hunk 3 above)
Comment (evancharlton) [reply src/components/ZoomRect/index.js]: Done with the latest commit. Please take a look (you only need to view changes from that diff if you want to)

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

File: src/components/ZoomRect/index.js (hunk 3 above)
Comment (danlevings) [reply src/components/ZoomRect/index.js]: SGTM 👍

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

File: src/components/ZoomRect/index.js [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+        const subDomain = (this.props.subDomainsByItemId[itemId] || {})[axis];
+        const subDomainRange = subDomain[1] - subDomain[0];
+        let newSubDomain = null;
+        if (sourceEvent.deltaY) {
+          // This is a zoom event.
+          const { deltaMode, deltaY } = sourceEvent;
+          const percentFromEnd = percents[axis];
+
+          // This was borrowed from d3-zoom.
+          const zoomFactor = (deltaY * (deltaMode ? 120 : 1)) / 500;
```
Comment (valerii-cognite) [inline src/components/ZoomRect/index.js]: oh, man

File: src/components/InteractionLayer/index.js [hunk 5]
```diff
@@ -479,39 +434,17 @@ class InteractionLayer extends React.Component {
     }
   };
 
-  zoomed = () => {
-    const { ruler, zoomMode, onZoomXAxis } = this.props;
-    if (ruler && ruler.visible) {
-      this.processMouseMove(this.state.touchX, this.state.touchY);
```
Comment (valerii-cognite) [inline src/components/InteractionLayer/index.js]: I believe this call is missed now, so the ruler does not stick to position while dragging.

File: src/components/ZoomRect/index.js [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    };
+    itemIds.forEach(itemId => {
+      updates[itemId] = {};
+      Axes.ALL.filter(axis => zoomAxes[axis] && this.lastDeltas[axis]).forEach(
+        axis => {
+          const subDomain = (subDomainsByItemId[itemId] || {})[axis];
+          const subDomainRange = subDomain[1] - subDomain[0];
+          // TODO: Find the center of the touches and then place that on the
+          // rect so that we can figure out where to zoom relative to.
+          const percentFromEnd = 0.5;
```
Comment (valerii-cognite) [inline src/components/ZoomRect/index.js]: This one is pretty crucial for OpSup

File: src/components/ZoomRect/index.js [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+        if (newSubDomain) {
+          updates[itemId][axis] = newSubDomain;
+        }
+      });
+    });
+    this.lastTouch = { ...newTouchPosition };
+    return updates;
+  };
+
+  performTouchZoom = (touches, distances) => {
```
Comment (valerii-cognite) [inline src/components/ZoomRect/index.js]: I'm a bit afraid of reimplementing this functionality of d3

File: stories/Scatterplot.stories.js [hunk 8]
```diff
@@ -488,4 +565,43 @@ storiesOf('Scatterplot', module)
...
+          pointsPerSeries={100}
+          series={[{ id: '1 2', color: 'steelblue' }]}
+          timeAccessor={d => +d.timestamp}
+          xAccessor={d => +d.x}
+          yAccessor={d => +d.y}
+        >
+          <div style={{ height: 500 }}>
+            <Scatterplot zoomable />
+          </div>
+          <div style={{ width: 450 }}>
```
[the remark this answers is not in the document]
Comment (evancharlton) [reply stories/Scatterplot.stories.js]: Agreed, the `<ContextChart>` item needs an API scrubbing before it "goes live" -- but I would prefer to do that in a separate PR. Sound good?

[the remark this answers is not in the document]
Comment (valerii-cognite) [reply stories/Scatterplot.stories.js]: sure, it is not related that match here

File: src/components/ZoomRect/index.js (hunk 7 above)
Comment (evancharlton) [reply src/components/ZoomRect/index.js]: I didn't want to do it, but there's basically no way around it as far as I can see. I fought d3 for a long time to get touch + mouse zoom working together, but they seem to only operate using transformations, which we basically can't use.

File: src/components/ZoomRect/index.js (hunk 6 above)
Comment (evancharlton) [reply src/components/ZoomRect/index.js]: I'll take a look at fixing this today.

File: src/components/ZoomRect/index.js (hunk 4 above)
Comment (evancharlton) [reply src/components/ZoomRect/index.js]: Agreed.

**Comment by evancharlton** [comment]
Done -- ready for re-review!

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

File: src/components/InteractionLayer/index.js (hunk 5 above)
Comment (evancharlton) [reply src/components/InteractionLayer/index.js]: I can't figure out what this actually does -- I think you'll need to show me in-person.

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

File: src/components/Scaler/index.js [hunk 9]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
-        yTransformations: {},
-      });
-      return;
-    }
+
     if (
-      nextPropsDomain[0] !== prevXDomain[0] ||
-      nextPropsDomain[1] !== prevXDomain[1]
+      nextTimeDomain[0] !== prevTimeDomain[0] ||
+      nextTimeDomain[1] !== prevTimeDomain[1]
```
Comment (jasmineleeson) [inline src/components/Scaler/index.js]: isn't there already a check to see if the previous time domain is different from the new one? can't we just move the logic inside this if statement into the other one?

File: src/components/Scaler/index.js [hunk 10]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
+              nextTimeDomain[1] - dt,
+              nextTimeDomain[1],
+            ];
+          } else if (timeSubDomain[0] <= prevTimeDomain[0]) {
+            // Looking at the back of the window -- continue to track that.
+            subDomainsByItemId[item.id][Axes.time] = [
+              prevTimeDomain[0],
+              prevTimeDomain[0] + dt,
+            ];
+          }
```
Comment (jasmineleeson) [inline src/components/Scaler/index.js]: I don't see the need to completely reconstruct `subDomainsByItemId` here. Couldn't you just iterate through the current state's  `subDomainsByItemId` and change what needs to be changed (like what you did above when the previous timeSubDomain was not equal to the new one)?

File: src/components/Scaler/index.js [hunk 11]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
-      ySubDomain: ySubDomains[c.id] || c.ySubDomain,
-    }));
-
     const enrichedContext = {
-      collections: enrichedCollections,
-      series: enrichedSeries,
-      xSubDomain: xSubDomain || dataContext.xSubDomain,
+      collections: dataContext.collections,
+      series: dataContext.series,
+      timeSubDomain: timeSubDomain || dataContext.timeSubDomain,
```
Comment (jasmineleeson) [inline src/components/Scaler/index.js]: I don't actually see the `timeSubDomain` state being used anywhere in Scaler other than here. In which case, can we just remove it from the state and have `dataContext.timeSubDomain || dataContext.timeDomain` here?

File: src/components/Scaler/index.js [hunk 12]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
-      ...c,
-      ySubDomain: ySubDomains[c.id] || c.ySubDomain,
-    }));
-
     const enrichedContext = {
-      collections: enrichedCollections,
-      series: enrichedSeries,
-      xSubDomain: xSubDomain || dataContext.xSubDomain,
+      collections: dataContext.collections,
+      series: dataContext.series,
```
Comment (jasmineleeson) [inline src/components/Scaler/index.js]: collections and series are not in enriched anymore so what is the point of them being here

File: src/components/InteractionLayer/index.js (hunk 5 above)
Comment (evancharlton) [reply src/components/InteractionLayer/index.js]: Done.

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

File: src/components/Scaler/index.js [hunk 13]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
+      domainsByItemId: {},
 
+      // Map from item (collection, series) to their respective subdomains.
+      subDomainsByItemId: {},
+    };
+    const domainsByItemId = this.getDomainsByItemId();
+    const subDomainsByItemId = this.getSubDomainsByItemId();
+
+    this.state.domainsByItemId = domainsByItemId;
+    this.state.subDomainsByItemId = subDomainsByItemId;
```
[the remark this answers is not in the document]
Comment (evancharlton) [reply src/components/Scaler/index.js]: Done.

File: src/components/Scaler/index.js [hunk 14]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
     if (
       !isEqual(
-        prevProps.dataContext.xSubDomain,
-        this.props.dataContext.xSubDomain
+        prevProps.dataContext.timeSubDomain,
+        this.props.dataContext.timeSubDomain
       )
     ) {
+      // When timeSubDomain changes, we need to update everything downstream.
+      const subDomainsByItemId = { ...this.state.subDomainsByItemId };
```
[the remark this answers is not in the document]
Comment (evancharlton) [reply src/components/Scaler/index.js]: I don't want to mutate `this.state`

File: src/components/Scaler/index.js [hunk 15]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
-        this.props.dataContext.externalXSubDomain
+        prevProps.dataContext.timeDomain,
+        this.props.dataContext.timeDomain
       )
     ) {
-      // eslint-disable-next-line react/no-did-update-set-state
-      this.setState({
-        xSubDomain: this.props.dataContext.externalXSubDomain,
+      // When timeDomain changes, we need to update everything downstream.
+      const domainsByItemId = { ...this.state.domainsByItemId };
```
[the remark this answers is not in the document]
Comment (evancharlton) [reply src/components/Scaler/index.js]: same as above

File: src/components/Scaler/index.js (hunk 9 above)
Comment (evancharlton) [reply src/components/Scaler/index.js]: Right you are! Done

File: src/components/Scaler/index.js (hunk 10 above)
Comment (evancharlton) [reply src/components/Scaler/index.js]: Yeah they're probably the same. (I was just copying the window-tracking logic from InteractionLayer over without following what it did)

File: src/components/Scaler/index.js [hunk 16]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
-        this.props.dataContext.xSubDomainChanged(xSubDomain);
-      }
-    );
+    const domainsByItemId = []
+      .concat(dataContext.series)
+      .concat(dataContext.collections)
+      .reduce(
+        (acc, item) => ({
+          ...acc,
+          [item.id]: this.state.domainsByItemId[item.id] || {
```
[the remark this answers is not in the document]
Comment (evancharlton) [reply src/components/Scaler/index.js]: Good catch!

File: src/components/Scaler/index.js [hunk 17]
```diff
@@ -26,201 +46,262 @@ class Scaler extends Component {
...
+            time: [...dataContext.timeDomain],
+            x: [...(item.xDomain || [])],
+            y: [...(item.yDomain || [])],
```
[the remark this answers is not in the document]
Comment (evancharlton) [reply src/components/Scaler/index.js]: Whoops!

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

File: src/components/ZoomRect/index.js [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+      this.rectSelection.on('touchmove', this.onTouchMove, true);
+      this.rectSelection.on('touchstart', this.onTouchStart, true);
+      this.rectSelection.call(this.zoom.on('zoom', this.zoomed));
+      this.rectSelection.on('dblclick.zoom', null);
+    } else {
+      this.rectSelection.on('.zoom', null);
+    }
+  };
+
+  zoomed = () => {
```
Comment (jasmineleeson) [inline src/components/ZoomRect/index.js]: is `zoomed` used for desktop zoom whereas `performTouchZoom` is used for mobile? If this is the case can you document that?

File: src/components/ZoomRect/index.js [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+    const deltas = {
+      [Axes.time]: Math.abs(touchOneX - touchTwoX),
+      [Axes.x]: Math.abs(touchOneX - touchTwoX),
+      [Axes.y]: Math.abs(touchOneY - touchTwoY),
+    };
+    const multipliers = {
+      [Axes.time]: -1,
+      [Axes.x]: -1,
+      [Axes.y]: 1,
+    };
```
[the remark this answers is not in the document]
Comment (evancharlton) [reply src/components/ZoomRect/index.js]: I'm gonna hate all of these fancy multipliers in the future, I just know it.

I really need to abstract this out somehow.

(Done)

File: src/components/ZoomRect/index.js (hunk 18 above)
Comment (evancharlton) [reply src/components/ZoomRect/index.js]: Sure.

File: src/components/Scaler/index.js (hunk 12 above)
Comment (evancharlton) [reply src/components/Scaler/index.js]: Good catch.

File: src/components/Scaler/index.js (hunk 11 above)
Comment (evancharlton) [reply src/components/Scaler/index.js]: Amazing find! Yep, there's no need to have it here.

**Comment by cognite-cicd** [comment]
[pr-server]
The storybook for this PR is hosted on https://griff-react-pr-191.surge.sh

**Pull request merged**